docs/blog/2021-06-16-tailor-your-slack-audit-sink-messages.md

```diff
@@ -0,0 +1,149 @@
+
+For each incoming API request into Clutch, an audit event is [created, persisted, and then forwarded](https://clutch.sh/docs/advanced/security-auditing/) to the sinks. The role of a sink is simple: process the audit event appropriately for your needs and forward the event to other systems (Clutch ships with a logging and Slack sink and you can add integration for new ones, such as Microsoft Teams). The sinks utilize the recorded information in an audit event to create an output. For example, the Slack sink’s [default message](https://github.com/lyft/clutch/blob/b56d9a929361fb590144bd8fd25d6b34d0fd5289/backend/service/auditsink/slack/slack.go#L114-L131) is created using the [event's](https://github.com/lyft/clutch/blob/b56d9a929361fb590144bd8fd25d6b34d0fd5289/api/audit/v1/audit.proto#L60-L76) `username`, `service_name`, `method_name`, and `resources` fields.
+
+The first version of the audit middleware (runs on each request) collected only a subset of information from the incoming API calls. In order to support richer output in the audit sinks (ex. customized Slack messages), the audit middleware needed to be modified to record the entire API request/response body, which serves as metadata in the audit event.
+
+## The Any
+
+Naturally request and response messages vary across APIs, and the audit middleware doesn’t know the message types in advance (i.e. is this message a [DescribePod](https://github.com/lyft/clutch/blob/a07d897dc69229ad1453fea6dcaf36ad50d2f6a6/api/k8s/v1/k8s.proto#L225-L235) request or a [CreateExperiment](https://github.com/lyft/clutch/blob/a07d897dc69229ad1453fea6dcaf36ad50d2f6a6/api/chaos/experimentation/v1/experimentation.proto#L21-L24) response?). Therefore, we used Protobuf's [Any](https://developers.google.com/protocol-buffers/docs/proto3#any) type to generically serialize the API request/response and save it in the audit event. The Any is composed of two values: the serialized message as bytes and a unique URL for identifying and resolving the message type.
+
+Modifications made to the audit event schema:
+
+```protobuf
+message RequestEvent {
+  ...
+  // The API request saved as metadata for the event.
+  RequestMetadata request_metadata = 7;
+  // The API response saved as metadata for the event.
+  ResponseMetadata response_metadata = 8;
+}
+
+message RequestMetadata {
+  google.protobuf.Any body = 1;
+}
+
+message ResponseMetadata {
+  google.protobuf.Any body = 1;
+}
+```
+
+The event’s Any fields can then be unpacked into its underlying message type by the audit sinks.
+
+Below is an example of an audit event with the stored metadata for [DescribePod](https://github.com/lyft/clutch/blob/a07d897dc69229ad1453fea6dcaf36ad50d2f6a6/api/k8s/v1/k8s.proto#L16-L22):
+
+```json
+{
+...
+"requestMetadata":{"body":{"@type":"type.googleapis.com/clutch.k8s.v1.DescribePodRequest","clientset":"kind-clutch-local","cluster":"kind-clutch-local","namespace":"envoy-staging","name":"envoy-58d647f457-5b2k4"}},
```

Comment:
can we wrap these? The text doesn't show well on the rendered page

Reply:
looked at following some of the suggestions here https://stackoverflow.com/questions/41238148/how-to-line-wrap-long-lines-inside-markdown-code-blocks-in-githubgitlab (in particular `white-space : pre-wrap ;`) but it still wasn't readable so I decided to not-minify the json. I think it's better now, lmk what you think